### DISCUSSION TITLE
About JBI offline program

### DISCUSSION THREAD

**Discussion by yfc26** [Yaskawa-Global/YMConnect#26] [General]
Hi, 
I create a JBI file by C++, and download the file to DX200 by the function LoadToControllerFromPath. But there are several error appears .
![err1](https://github.com/user-attachments/assets/25502a5e-e4a1-462d-9fdf-13c77fb48208)
![err2](https://github.com/user-attachments/assets/b6316608-d794-4129-b1e8-756de7049327)
![err3](https://github.com/user-attachments/assets/21bff6b5-1d1d-4175-9c08-4eddc0c4f2a1)

Can you tell me how to program an offline JBI file? Thank you very much.
The code is also attached.
[code.txt](https://github.com/user-attachments/files/17596961/code.txt)

**Comment by aseamans** [comment]
Hi @yfc26,
In your job, The line endings need to be formatted as "\r\n"
Try changing this line of code.
`std::string theCRLF = "\n";  `
to
`std::string theCRLF = "\r\n"; `
and let us know what happens.

**Comment by yfc26** [comment]
Hi @aseamans , 
I tried again. It can work with "\n" now. I met another problem. When I used "  thefile << "thefile << "///POSTYPE PULSE" + theCRLF;    thefile << "///PULSE" + theCRLF; ", I can download the file into robot, but when I changed this to"thefile << "thefile << "///POSTYPE ROBOT" + theCRLF;    thefile << "///RECTAN" + theCRLF; ", that is, changing the pulse to cartesian with all other codes not changed, an error occured as, 
![err](https://github.com/user-attachments/assets/51995f59-f89e-418d-9c93-aace08811809)
Can you tell me what's wrong? Thanks a lot.

**Reply by ted-miller** [reply]
@yfc26 Could you please post an unedited copy of the final job that your application generates?

Generally speaking, the specification for the job header format is detailed in the `Relative Job` user manual. There is a link to that document at the top of this article: https://knowledge.motoman.com/hc/en-us/articles/4407113150231-Maximizing-Your-Robot-s-Potential-with-Relative-Job

**Comment by yfc26** [comment]
Hi @ted-miller , please find the attached final job. Thank you for your help.
/JOB
//NAME ABC
//POS
///NPOS 2,0,0,0,0,0
///TOOL 0
///POSTYPE ROBOT
///RECTAN
C00000=200.1,250.2,300.3,40.4,50.5,60.6
C00001=100.1,200.2,300.3,40.4,50.5,60.6
//INST
///DATE 2024/11/05 08:38
///ATTR SC,RW
///GROUP1 RB1
NOP
DOUT OT#(1) ON
SET B000 0
MOVJ C00000 VJ=25.00
MOVL C00001  V=200.0
SET B000 2
DOUT OT#(1) OFF
END

**Reply by ted-miller** [reply]
There are two issues with this job.

1) For cartesian coordinates, the XYZ positions should have 3 decimal places. The rotation values should have 4 decimal places.

2) For each cartesian point, you must specify the form/figure/posture/closure (I've heard it called any of those terms). For example:

```
/JOB
//NAME XYZ
//POS
///NPOS 2,0,0,0,0,0
///TOOL 1
///POSTYPE ROBOT
///RECTAN
///RCONF 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
C00000=1245.000,0.000,960.000,180.0000,-90.0000,0.0000
///RCONF 1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
C00001=1386.602,0.000,959.993,180.0000,-89.9996,0.0000
//INST
///DATE 2022/07/14 09:44
///ATTR SC,RW,RJ
////FRAME ROBOT
///GROUP1 RB1
NOP
MOVJ C00000 VJ=25.00
MOVL C00001 V=200.0
END

```

The definition of the `RCONF` header is described in chapter 4 of document 178659-1CD, which is available here: https://www.motoman.com/en-us/service-training/product-documentation

**Comment by yfc26** [comment]
@ted-miller 
Thank you for your great help. I'll try it.
